README.md

```diff
@@ -174,115 +176,129 @@ lego defaults to communicating with the production Let's Encrypt ACME server. If
 lego --server=https://acme-staging-v02.api.letsencrypt.org/directory …
 ```
 
-## DNS Challenge API Details
-
-### AWS Route 53
-
-The following AWS IAM policy document describes the permissions required for lego to complete the DNS challenge.
-Replace `<INSERT_YOUR_HOSTED_ZONE_ID_HERE>` with the Route 53 zone ID of the domain you are authorizing.
-
-```json
-{
-    "Version": "2012-10-17",
-    "Statement": [
-        {
-            "Effect": "Allow",
-            "Action": [
-                "route53:GetChange",
-                "route53:ListHostedZonesByName"
-            ],
-            "Resource": [
-                "*"
-            ]
-        },
-        {
-            "Effect": "Allow",
-            "Action": [
-                "route53:ChangeResourceRecordSets"
-            ],
-            "Resource": [
-                "arn:aws:route53:::hostedzone/<INSERT_YOUR_HOSTED_ZONE_ID_HERE>"
-            ]
-        }
-    ]
-}
-```
-
 ## ACME Library Usage
 
 A valid, but bare-bones example use of the acme package:
 
 ```go
+package sample
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"log"
+
+	"github.com/xenolf/lego/acme"
+	"github.com/xenolf/lego/certcrypto"
+	"github.com/xenolf/lego/registration"
+)
+
 // You'll need a user or account type that implements acme.User
 type MyUser struct {
 	Email        string
-	Registration *acme.RegistrationResource
+	Registration *registration.Resource
 	key          crypto.PrivateKey
 }
-func (u MyUser) GetEmail() string {
+
+func (u *MyUser) GetEmail() string {
 	return u.Email
 }
-func (u MyUser) GetRegistration() *acme.RegistrationResource {
+func (u MyUser) GetRegistration() *registration.Resource {
 	return u.Registration
-}
-func (u MyUser) GetPrivateKey() crypto.PrivateKey {
+}*
+func (u *MyUser) GetPrivateKey() crypto.PrivateKey {
 	return u.key
 }
 
-// Create a user. New accounts need an email and private key to start.
-const rsaKeySize = 2048
-privateKey, err := rsa.GenerateKey(rand.Reader, rsaKeySize)
-if err != nil {
-	log.Fatal(err)
-}
-myUser := MyUser{
-	Email: "you@yours.com",
-	key: privateKey,
-}
-
-// A client facilitates communication with the CA server. This CA URL is
-// configured for a local dev instance of Boulder running in Docker in a VM.
-client, err := acme.NewClient("http://192.168.99.100:4000/directory", &myUser, acme.RSA2048)
-if err != nil {
-  log.Fatal(err)
+func main() {
+
+	// Create a user. New accounts need an email and private key to start.
+	const rsaKeySize = 2048
+	privateKey, err := rsa.GenerateKey(rand.Reader, rsaKeySize)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	myUser := MyUser{
+		Email: "you@yours.com",
+		key:   privateKey,
+	}
+
+	// A client facilitates communication with the CA server. This CA URL is
+	// configured for a local dev instance of Boulder running in Docker in a VM.
+
+	config := acme.NewDefaultConfig(&myUser).
+		WithCADirURL("http://192.168.99.100:4000/directory").WithKeyType(certcrypto.RSA2048)
```

Comment:
I'm not a huge fan of these drive-thru configurations that read like a DSL. There's no way to return errors here, which makes me a little nervous because surely there's a way to pass in a bad configuration?

After inspecting the value that gets returned, I see some effort to keep certain fields unexported. Maybe they could just be exported instead? It's less weird to just do something like this:

```go
config := acme.NewDefaultConfig(&myUser)
config.CADirURL = "..."
config.KeyType = "..."
```

Reply:
The fluent syntax or the exported field way have the same problem if you want to manage errors.

I propose another way:
```go
config, err := acme.NewConfig(&myUser, acme.WithCADirURL("http://192.168.99.100:4000/directory"), acme.WithKeyType(certcrypto.RSA2048))
```

Reply:
Why not? Users need to set those values, they're literally supposed to be user-facing.

I have found that usually trying to keep fields unexported lends itself to more complexity for little or no real benefit.

Reply:
The configuration manages the configuration: the `Client` doesn't need to validate the configuration because it's not his responsibility (separation of concerns).

The lib's user must use `NewConfig`: if the fields are exported the user can easily create an invalid configuration because he doesn't need to call `NewConfig`.

The configuration must be used only inside the `client` package.

Thanks to the non-exported fields, we can easily modify the configuration struct without breaking the lib's user.

Reply:
I see. Well, those are nice benefits. They are needed from the perspective that an invalid config should not be allowed to be created in the first place. (I disagree personally, because functions should validate their input -- and since `NewClient` gets input, it should validate it. What's the harm in using it outside the `client` package? No other packages will probably use it, and if they do, so be it?)

Anyway, just my $0.02 from my own experience, where unexported fields have come back to bite more than actually help. But feel free proceed.

Reply:
I try to emulate the designs of the standard library as much as I can, so that new developers coming to the space (and old ones) know what's going on right away. Take a look at for example, the [httputil.ReverseProxy](https://golang.org/pkg/net/http/httputil/#ReverseProxy), which does quite a lot. It allows lots of configuration through it's exported fields. I feel if we are concerned that an internal component/field shouldn't be modified, or we plan to do so, it should be unexported instead, rather than creating a series of "initializer" style config functions, which I think are just a pretty wrapper around accessing the struct directly.

Reply:
I know well `httputil.ReverseProxy` and I'm not sure it can be use as a model for Lego.
However, if I am alone in seeing the benefits, I will not impose my point of view on this section.